In [1]:
import pandas as pd

In [ ]:
# Load the flight data and weather data
flight_data_path = "data/YVR_flight_data_2023-2024_departure.csv"
weather_data_path = "data/Weather-VAN21-24.csv"

In [3]:
# Read the CSV files
flight_data = pd.read_csv(flight_data_path)
weather_data = pd.read_csv(weather_data_path)

In [4]:
flight_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 281714 entries, 0 to 281713
Data columns (total 35 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   type                          281712 non-null  object 
 1   status                        281712 non-null  object 
 2   departure.iataCode            281712 non-null  object 
 3   departure.icaoCode            281712 non-null  object 
 4   departure.terminal            270859 non-null  object 
 5   departure.gate                260640 non-null  object 
 6   departure.delay               262261 non-null  float64
 7   departure.scheduledTime       281712 non-null  object 
 8   departure.estimatedTime       268697 non-null  object 
 9   arrival.iataCode              281600 non-null  object 
 10  arrival.icaoCode              281570 non-null  object 
 11  arrival.baggage               126982 non-null  object 
 12  arrival.gate                  211597 non-nul

In [5]:
weather_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26832 entries, 0 to 26831
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   time                  26832 non-null  object 
 1   temperature_2m        26832 non-null  float64
 2   relative_humidity_2m  26832 non-null  int64  
 3   apparent_temperature  26832 non-null  float64
 4   precipitation         26832 non-null  float64
 5   rain                  26832 non-null  float64
 6   snowfall              26832 non-null  float64
 7   weather_code          26832 non-null  int64  
 8   pressure_msl          26832 non-null  float64
 9   surface_pressure      26832 non-null  float64
 10  cloud_cover           26832 non-null  int64  
 11  wind_speed_10m        26832 non-null  float64
 12  wind_direction_10m    26832 non-null  int64  
 13  wind_gusts_10m        26832 non-null  float64
dtypes: float64(9), int64(4), object(1)
memory usage: 2.9+ MB


In [6]:
# Ensure that the columns are in datetime format
flight_columns = ['departure.scheduledTime','departure.estimatedTime','arrival.scheduledTime','departure.actualTime','arrival.estimatedTime','arrival.actualTime']

for col in flight_columns:
    flight_data[col] = pd.to_datetime(flight_data[col], errors='coerce')

weather_data['time'] = pd.to_datetime(weather_data['time'], errors='coerce')

In [7]:
flight_data.head()

,type,status,departure.iataCode,departure.icaoCode,departure.terminal,departure.gate,departure.delay,departure.scheduledTime,departure.estimatedTime,arrival.iataCode,...,codeshared.flight.icaoNumber,departure.actualTime,departure.estimatedRunway,departure.actualRunway,arrival.terminal,arrival.delay,arrival.estimatedTime,arrival.actualTime,arrival.estimatedRunway,arrival.actualRunway
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaT,NaN,...,NaN,NaT,NaN,NaN,NaN,NaN,NaT,NaT,NaN,NaN
1,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,wja614,NaT,NaN,NaN,NaN,NaN,NaT,NaT,NaN,NaN
2,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,wja614,NaT,NaN,NaN,NaN,NaN,NaT,NaT,NaN,NaN
3,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,wja614,NaT,NaN,NaN,NaN,NaN,NaT,NaT,NaN,NaN
4,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,wja614,NaT,NaN,NaN,NaN,NaN,NaT,NaT,NaN,NaN


In [8]:
weather_data.head()

,time,temperature_2m,relative_humidity_2m,apparent_temperature,precipitation,rain,snowfall,weather_code,pressure_msl,surface_pressure,cloud_cover,wind_speed_10m,wind_direction_10m,wind_gusts_10m
0,2021-08-31 00:00:00,17.1,53,14.2,0.0,0.0,0.0,1,1012.1,1003.4,42,15.7,171,28.1
1,2021-08-31 01:00:00,16.8,54,14.8,0.0,0.0,0.0,3,1011.9,1003.2,100,9.3,164,27.4
2,2021-08-31 02:00:00,16.4,55,15.6,0.0,0.0,0.0,3,1011.7,1003.0,100,0.8,333,17.6
3,2021-08-31 03:00:00,15.7,60,14.3,0.0,0.0,0.0,2,1012.0,1003.3,79,6.3,301,10.4
4,2021-08-31 04:00:00,14.8,68,13.3,0.1,0.1,0.0,51,1012.5,1003.8,63,8.3,304,15.5


In [9]:
# Get the minimum and maximum dates directly from the datetime columns
flight_date_range = (flight_data['departure.scheduledTime'].min(), flight_data['departure.scheduledTime'].max())
weather_date_range = (weather_data['time'].min(), weather_data['time'].max())

print(f"Flight data date range: {flight_date_range}")
print(f"Weather data date range: {weather_date_range}")

Flight data date range: (Timestamp('2023-10-25 17:00:00'), Timestamp('2024-09-25 16:55:00'))
Weather data date range: (Timestamp('2021-08-31 00:00:00'), Timestamp('2024-09-21 23:00:00'))


In [10]:
# Define the date range for filtering
start_date = pd.Timestamp('2023-10-25 17:00:00')
end_date = pd.Timestamp('2024-09-21 23:00:00')

# Filter the weather data and flight data to include only rows within the specified date range
filtered_flight_data = flight_data[(flight_data['departure.scheduledTime'] >= start_date) & (flight_data['departure.scheduledTime'] <= end_date)]
filtered_weather_data = weather_data[(weather_data['time'] >= start_date) & (weather_data['time'] <= end_date)]

In [11]:
# Check for any non-datetime values (NaT) in both columns
print(f"Number of invalid datetime entries in flight data: {filtered_flight_data['departure.scheduledTime'].isna().sum()}")
print(f"Number of invalid datetime entries in weather data: {filtered_weather_data['time'].isna().sum()}")

# Display a few samples to confirm the datetime format
print(filtered_flight_data['departure.scheduledTime'].head())
print(filtered_weather_data['time'].head())

Number of invalid datetime entries in flight data: 0
Number of invalid datetime entries in weather data: 0
1   2023-10-25 17:00:00
2   2023-10-25 17:00:00
3   2023-10-25 17:00:00
4   2023-10-25 17:00:00
5   2023-10-25 17:00:00
Name: departure.scheduledTime, dtype: datetime64[ns]
18857   2023-10-25 17:00:00
18858   2023-10-25 18:00:00
18859   2023-10-25 19:00:00
18860   2023-10-25 20:00:00
18861   2023-10-25 21:00:00
Name: time, dtype: datetime64[ns]


In [12]:
# Convert the datetime columns to the desired format (YYYY-MM-DD HH)
filtered_flight_data['departure.scheduledTime2'] = pd.to_datetime(filtered_flight_data['departure.scheduledTime'], errors='coerce').dt.strftime('%Y-%m-%d %H')
filtered_weather_data['time2'] = pd.to_datetime(weather_data['time'], errors='coerce').dt.strftime('%Y-%m-%d %H')

C:\Users\sanda\AppData\Local\Temp\ipykernel_29428\3359587195.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_flight_data['departure.scheduledTime2'] = pd.to_datetime(filtered_flight_data['departure.scheduledTime'], errors='coerce').dt.strftime('%Y-%m-%d %H')
C:\Users\sanda\AppData\Local\Temp\ipykernel_29428\3359587195.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_weather_data['time2'] = pd.to_datetime(weather_data['time'], errors='coerce').dt.strftime('%Y-%m-%d %H')


In [13]:
# Perform the merge using the common datetime columns
merged_data = pd.merge(filtered_flight_data, filtered_weather_data, left_on='departure.scheduledTime2', right_on='time2', how='left')

# Display the number of rows in the merged dataset and show the first few rows
print("Number of rows in the merged dataset:", len(merged_data))
merged_data.head()

Number of rows in the merged dataset: 278412


,type,status,departure.iataCode,departure.icaoCode,departure.terminal,departure.gate,departure.delay,departure.scheduledTime,departure.estimatedTime,arrival.iataCode,...,rain,snowfall,weather_code,pressure_msl,surface_pressure,cloud_cover,wind_speed_10m,wind_direction_10m,wind_gusts_10m,time2
0,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,0.0,0.0,2,1013.1,1004.1,58,4.9,234,10.8,2023-10-25 17
1,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,0.0,0.0,2,1013.1,1004.1,58,4.9,234,10.8,2023-10-25 17
2,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,0.0,0.0,2,1013.1,1004.1,58,4.9,234,10.8,2023-10-25 17
3,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,0.0,0.0,2,1013.1,1004.1,58,4.9,234,10.8,2023-10-25 17
4,departure,active,yvr,cyvr,d,b15,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,yow,...,0.0,0.0,2,1013.1,1004.1,58,4.9,234,10.8,2023-10-25 17


In [ ]:
# Save the merged dataset to a CSV file
merged_data.to_csv('data/merged_flight_weather_data.csv', index=False)

In [ ]:
# Load the merged dataset
flight_weather_data = pd.read_csv("data/merged_flight_weather_data.csv")

Key Columns to Consider

Flight Information:

* departure.delay: Indicates if a delay occurred and how long it was.
* departure.scheduledTime: Original scheduled time for departure.
* departure.estimatedTime: Adjusted estimated time of departure.
* departure.actualTime: Actual departure time.
* status: Current status of the flight (e.g., delayed, on-time).
* airline.name: The airline could help identify patterns of delays with specific airlines.

Weather Information:

* temperature_2m: Extreme temperatures can affect flight operations.
* precipitation, rain, and snowfall: These factors are directly related to adverse weather conditions that can cause delays.
* wind_speed_10m and wind_gusts_10m: High wind speeds can lead to delays or cancellations.
* cloud_cover: Low visibility conditions might result in flight delays.
* weather_code: Encodes specific weather conditions (e.g., thunderstorms, fog).
* pressure_msl and surface_pressure: Rapid changes in pressure can indicate turbulent weather.
* relative_humidity_2m: High humidity can be related to foggy or stormy conditions.
* wind_direction_10m: Useful for understanding crosswinds that may affect takeoff and landing.

In [16]:
# Filter the necessary columns
filtered_data = flight_weather_data[
    [
        "status", 
        "departure.delay", 
        "departure.scheduledTime", 
        "departure.estimatedTime", 
        "departure.actualTime", 
        "airline.name", 
        "temperature_2m", 
        "precipitation", 
        "rain", 
        "snowfall", 
        "weather_code", 
        "cloud_cover", 
        "wind_speed_10m", 
        "wind_gusts_10m", 
        "pressure_msl", 
        "surface_pressure", 
        "relative_humidity_2m",
        "wind_direction_10m"
    ]
]

In [ ]:
# Save the filtered data to a CSV
filtered_data.to_csv("data/final_flight_weather_data.csv", index=False)

In [ ]:
# Load your dataset 
flight_weather_data = pd.read_csv("data/final_flight_weather_data.csv")

In [19]:
flight_weather_data.head()

,status,departure.delay,departure.scheduledTime,departure.estimatedTime,departure.actualTime,airline.name,temperature_2m,precipitation,rain,snowfall,weather_code,cloud_cover,wind_speed_10m,wind_gusts_10m,pressure_msl,surface_pressure,relative_humidity_2m,wind_direction_10m
0,active,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,NaN,air france,5.6,0.0,0.0,0.0,2,58,4.9,10.8,1013.1,1004.1,72,234
1,active,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,NaN,jet linx aviation,5.6,0.0,0.0,0.0,2,58,4.9,10.8,1013.1,1004.1,72,234
2,active,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,NaN,korean air,5.6,0.0,0.0,0.0,2,58,4.9,10.8,1013.1,1004.1,72,234
3,active,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,NaN,qantas,5.6,0.0,0.0,0.0,2,58,4.9,10.8,1013.1,1004.1,72,234
4,active,53.0,2023-10-25 17:00:00,2023-10-25 17:53:00,NaN,westjet,5.6,0.0,0.0,0.0,2,58,4.9,10.8,1013.1,1004.1,72,234


In [20]:
flight_weather_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278412 entries, 0 to 278411
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   status                   278412 non-null  object 
 1   departure.delay          259163 non-null  float64
 2   departure.scheduledTime  278412 non-null  object 
 3   departure.estimatedTime  265508 non-null  object 
 4   departure.actualTime     212643 non-null  object 
 5   airline.name             278412 non-null  object 
 6   temperature_2m           278412 non-null  float64
 7   precipitation            278412 non-null  float64
 8   rain                     278412 non-null  float64
 9   snowfall                 278412 non-null  float64
 10  weather_code             278412 non-null  int64  
 11  cloud_cover              278412 non-null  int64  
 12  wind_speed_10m           278412 non-null  float64
 13  wind_gusts_10m           278412 non-null  float64
 14  pres

In [21]:
flight_weather_data.describe()

,departure.delay,temperature_2m,precipitation,rain,snowfall,weather_code,cloud_cover,wind_speed_10m,wind_gusts_10m,pressure_msl,surface_pressure,relative_humidity_2m,wind_direction_10m
count,259163.000000,278412.000000,278412.000000,278412.000000,278412.000000,278412.000000,278412.000000,278412.000000,278412.000000,278412.000000,278412.000000,278412.000000,278412.000000
mean,28.152012,10.057804,0.189962,0.179525,0.007306,14.146879,52.861795,9.810236,17.920604,1016.223967,1007.317798,81.933487,169.686181
std,35.682707,6.392863,0.580001,0.565182,0.090745,23.167151,40.821915,5.831865,9.796155,7.133256,7.066822,13.057460,98.065842
min,1.000000,-14.100000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.200000,991.300000,982.300000,18.000000,4.000000
25%,13.000000,5.400000,0.000000,0.000000,0.000000,0.000000,8.000000,5.400000,10.100000,1012.500000,1003.700000,74.000000,86.000000
50%,20.000000,9.300000,0.000000,0.000000,0.000000,2.000000,52.000000,8.700000,16.600000,1016.700000,1007.800000,85.000000,127.000000
75%,33.000000,15.000000,0.000000,0.000000,0.000000,3.000000,100.000000,13.400000,24.100000,1020.700000,1011.800000,92.000000,276.000000
max,1570.000000,29.100000,8.700000,8.700000,2.520000,75.000000,100.000000,42.800000,68.000000,1035.400000,1026.100000,100.000000,360.000000


In [22]:
# Check null counts for each column
flight_weather_data.isnull().sum()

status                         0
departure.delay            19249
departure.scheduledTime        0
departure.estimatedTime    12904
departure.actualTime       65769
airline.name                   0
temperature_2m                 0
precipitation                  0
rain                           0
snowfall                       0
weather_code                   0
cloud_cover                    0
wind_speed_10m                 0
wind_gusts_10m                 0
pressure_msl                   0
surface_pressure               0
relative_humidity_2m           0
wind_direction_10m             0
dtype: int64

Since this dataset has 278412 entries which is a huge count we decided to get the subset of data (by season)

In [23]:
# Convert 'departure.scheduledTime' to datetime if it isn't already
flight_weather_data['departure.scheduledTime'] = pd.to_datetime(flight_weather_data['departure.scheduledTime'], errors='coerce')

# Define a function to map months to seasons
def get_season(month):
    if month in [12, 1, 2]:
        return 'Winter'
    elif month in [3, 4, 5]:
        return 'Spring'
    elif month in [6, 7, 8]:
        return 'Summer'
    else:
        return 'Fall'

# Create the 'season' column based on 'departure.scheduledTime'
flight_weather_data['season'] = flight_weather_data['departure.scheduledTime'].dt.month.apply(get_season)

In [24]:
# Initialize an empty dictionary to store null counts by season
null_counts_by_season = {}

# Calculate null counts for each season and add to the dictionary
for season in ['Winter', 'Spring', 'Summer', 'Fall']:
    season_data = flight_weather_data[flight_weather_data['season'] == season]
    null_counts_by_season[season] = season_data.isnull().sum()

# Convert the dictionary to a DataFrame for a table structure
null_counts_df = pd.DataFrame(null_counts_by_season)

# Display the null counts table
print("Null counts by season:")
print(null_counts_df)

Null counts by season:
                         Winter  Spring  Summer   Fall
status                        0       0       0      0
departure.delay            4662    5206    6099   3282
departure.scheduledTime       0       0       0      0
departure.estimatedTime    3202    2906    3726   3070
departure.actualTime      19306   14455   20879  11129
airline.name                  0       0       0      0
temperature_2m                0       0       0      0
precipitation                 0       0       0      0
rain                          0       0       0      0
snowfall                      0       0       0      0
weather_code                  0       0       0      0
cloud_cover                   0       0       0      0
wind_speed_10m                0       0       0      0
wind_gusts_10m                0       0       0      0
pressure_msl                  0       0       0      0
surface_pressure              0       0       0      0
relative_humidity_2m          0       0   

I decided to use Winter season filtered dataset.

In [ ]:
# Filter for Winter season data only
winter_data = flight_weather_data[flight_weather_data['season'] == 'Winter']

# Save the Winter season data to a CSV file
winter_data.to_csv('data/winter_season_data.csv', index=False)

In [ ]:
winter_season_data = pd.read_csv("data/winter_season_data.csv")
winter_season_data.isnull().sum()

status                         0
departure.delay             4662
departure.scheduledTime        0
departure.estimatedTime     3202
departure.actualTime       19306
airline.name                   0
temperature_2m                 0
precipitation                  0
rain                           0
snowfall                       0
weather_code                   0
cloud_cover                    0
wind_speed_10m                 0
wind_gusts_10m                 0
pressure_msl                   0
surface_pressure               0
relative_humidity_2m           0
wind_direction_10m             0
season                         0
dtype: int64

### Handling missing values

For each record, if departure.actualTime is missing but departure.delay is available, the departure.actualTime is estimated by adding departure.delay time to the departure.scheduledTime.

In [27]:
# Count rows where 'departure.actualTime' is null and 'departure.delay' is not null
count_null_actual_delay_present = winter_season_data[(winter_season_data['departure.actualTime'].isnull()) & 
                                                   (winter_season_data['departure.delay'].notnull())].shape[0]

print("Count of rows where 'departure.actualTime' is null and 'departure.delay' is not null:", count_null_actual_delay_present)

Count of rows where 'departure.actualTime' is null and 'departure.delay' is not null: 15645


In [28]:
# Convert 'departure.scheduledTime' to datetime format for calculations
winter_season_data['departure.scheduledTime'] = pd.to_datetime(winter_season_data['departure.scheduledTime'], errors='coerce')

# Check for null in 'departure.actualTime' and handle it based on 'departure.delay'
winter_season_data['departure.actualTime'] = winter_season_data.apply(
    lambda row: row['departure.scheduledTime'] + pd.Timedelta(minutes=row['departure.delay'])
    if pd.isnull(row['departure.actualTime']) and not pd.isnull(row['departure.delay'])
    else row['departure.actualTime'],
    axis=1
)

winter_season_data.isnull().sum()

status                        0
departure.delay            4662
departure.scheduledTime       0
departure.estimatedTime    3202
departure.actualTime       3661
airline.name                  0
temperature_2m                0
precipitation                 0
rain                          0
snowfall                      0
weather_code                  0
cloud_cover                   0
wind_speed_10m                0
wind_gusts_10m                0
pressure_msl                  0
surface_pressure              0
relative_humidity_2m          0
wind_direction_10m            0
season                        0
dtype: int64

Next we can check for each record, if departure.delay is missing but departure.actualTime is available, so the departure.delay is estimated by substracting departure.scheduledTime from the departure.actualTime time.

In [29]:
# Count rows where 'departure.delay' is null but 'departure.actualTime' is not null
count_null_delay_actual_present = winter_season_data[(winter_season_data['departure.delay'].isnull()) & 
                                                   (winter_season_data['departure.actualTime'].notnull())].shape[0]

print("Count of rows where 'departure.delay' is null but 'departure.actualTime' is available:", count_null_delay_actual_present)

Count of rows where 'departure.delay' is null but 'departure.actualTime' is available: 1001


Before applying the condition, I checked for the condition of where departure.actualTime is greater than departure.scheduledTime and if the condition apply then only impute the departure delay is by substracting departure.scheduledTime from the departure.actualTime time.

In [30]:
# Ensure that 'departure.actualTime' and 'departure.scheduledTime' are converted to datetime
winter_season_data['departure.actualTime'] = pd.to_datetime(winter_season_data['departure.actualTime'], errors='coerce')
winter_season_data['departure.scheduledTime'] = pd.to_datetime(winter_season_data['departure.scheduledTime'], errors='coerce')

# Check for rows where 'departure.delay' is null, 'departure.actualTime' is not null, and 'departure.actualTime' is greater than 'departure.scheduledTime'
condition_1 = winter_season_data['departure.delay'].isnull()
condition_2 = winter_season_data['departure.actualTime'].notnull()
condition_3 = winter_season_data['departure.actualTime'] > winter_season_data['departure.scheduledTime']

# Combine the conditions
combined_condition = condition_1 & condition_2 & condition_3

# Check the count of records where this condition is met
count_conditions_met = winter_season_data[combined_condition].shape[0]
print(f"Count of records where 'departure.actualTime' is greater than 'departure.scheduledTime' "
      f"and 'departure.delay' is missing but 'departure.actualTime' is available: {count_conditions_met}")

# Handle the rows where conditions are met
winter_season_data.loc[combined_condition, 'departure.delay'] = (
    (winter_season_data['departure.actualTime'] - winter_season_data['departure.scheduledTime'])
    .dt.total_seconds() / 60
)

Count of records where 'departure.actualTime' is greater than 'departure.scheduledTime' and 'departure.delay' is missing but 'departure.actualTime' is available: 0


Then, I checked for the condition of where departure.actualTime is less than departure.scheduledTime and if the condition apply then remove the records, assuming that flights will not depart before scheduled time.

In [31]:
condition_4 = winter_season_data['departure.actualTime'] < winter_season_data['departure.scheduledTime']

combined_condition1 = condition_1 & condition_2 & condition_4

count_conditions_met1 = winter_season_data[combined_condition1].shape[0]
print(f"Count of records where 'departure.actualTime' is less than 'departure.scheduledTime' "
      f"and 'departure.delay' is missing but 'departure.actualTime' is available: {count_conditions_met1}")

# Remove these records from the dataset
winter_season_data = winter_season_data[~combined_condition1]

Count of records where 'departure.actualTime' is less than 'departure.scheduledTime' and 'departure.delay' is missing but 'departure.actualTime' is available: 847


In [32]:
winter_season_data.isnull().sum()

status                        0
departure.delay            3815
departure.scheduledTime       0
departure.estimatedTime    3103
departure.actualTime       3661
airline.name                  0
temperature_2m                0
precipitation                 0
rain                          0
snowfall                      0
weather_code                  0
cloud_cover                   0
wind_speed_10m                0
wind_gusts_10m                0
pressure_msl                  0
surface_pressure              0
relative_humidity_2m          0
wind_direction_10m            0
season                        0
dtype: int64

Then, I would like to check if same records are missing the both departure.delay and departure.actualTime constraint.

In [33]:
# Check if the same records have missing values in both 'departure.delay' and 'departure.actualTime'
missing_both = winter_season_data[winter_season_data['departure.delay'].isnull() & winter_season_data['departure.actualTime'].isnull()]

# Count of rows where both 'departure.delay' and 'departure.actualTime' are missing
print("Number of records missing both 'departure.delay' and 'departure.actualTime':", missing_both.shape[0])

Number of records missing both 'departure.delay' and 'departure.actualTime': 3661


There are 3661 records with 'departure.delay' and 'departure.actualTime' is missing. We are going to assume that there were no delays in that flights, and replace missing values in 'departure.delay' as 0 while 'departure.actualTime' to be the exact time as departure.scheduledTime

In [34]:
# Identify rows where both 'departure.delay' and 'departure.actualTime' are missing
missing_indices = winter_season_data[
    winter_season_data['departure.delay'].isnull() & winter_season_data['departure.actualTime'].isnull()
].index

# Set 'departure.delay' to 0 for these rows
winter_season_data.loc[missing_indices, 'departure.delay'] = 0

# Set 'departure.actualTime' to the values from 'departure.scheduledTime' for these rows
winter_season_data.loc[missing_indices, 'departure.actualTime'] = winter_season_data.loc[missing_indices, 'departure.scheduledTime']

winter_season_data.isnull().sum()

status                        0
departure.delay             154
departure.scheduledTime       0
departure.estimatedTime    3103
departure.actualTime          0
airline.name                  0
temperature_2m                0
precipitation                 0
rain                          0
snowfall                      0
weather_code                  0
cloud_cover                   0
wind_speed_10m                0
wind_gusts_10m                0
pressure_msl                  0
surface_pressure              0
relative_humidity_2m          0
wind_direction_10m            0
season                        0
dtype: int64

We have missing values in departure.estimatedTime count of 3103. Assuming that the flights departed on the departure.scheduledTime I am going to impute the missing values in estimatedTime with scheduledTime. 

In [35]:
# Impute missing values in 'departure.estimatedTime' with values from 'departure.actualTime'
winter_season_data['departure.estimatedTime'] = winter_season_data['departure.estimatedTime'].fillna(winter_season_data['departure.scheduledTime'])

winter_season_data.isnull().sum()

status                       0
departure.delay            154
departure.scheduledTime      0
departure.estimatedTime      0
departure.actualTime         0
airline.name                 0
temperature_2m               0
precipitation                0
rain                         0
snowfall                     0
weather_code                 0
cloud_cover                  0
wind_speed_10m               0
wind_gusts_10m               0
pressure_msl                 0
surface_pressure             0
relative_humidity_2m         0
wind_direction_10m           0
season                       0
dtype: int64

In [36]:
# Impute missing 'departure.delay' values where 'departure.actualTime' is available
winter_season_data['departure.delay'] = winter_season_data.apply(
    lambda row: (row['departure.actualTime'] - row['departure.scheduledTime']).total_seconds() / 60
    if pd.isnull(row['departure.delay']) and not pd.isnull(row['departure.actualTime']) 
    else row['departure.delay'],
    axis=1
)

winter_season_data.isnull().sum()

status                     0
departure.delay            0
departure.scheduledTime    0
departure.estimatedTime    0
departure.actualTime       0
airline.name               0
temperature_2m             0
precipitation              0
rain                       0
snowfall                   0
weather_code               0
cloud_cover                0
wind_speed_10m             0
wind_gusts_10m             0
pressure_msl               0
surface_pressure           0
relative_humidity_2m       0
wind_direction_10m         0
season                     0
dtype: int64

In [37]:
# Drop the 'season' column
winter_season_data = winter_season_data.drop(columns=['season'])

winter_season_data.shape

(68968, 18)

In our final flight weather dataset for fall season we have 47005 observations and 18 columns with no missing values.

In [ ]:
# Save the DataFrame to a CSV file
winter_season_data.to_csv('data/final_winter_season_data.csv', index=False)

In [ ]:
final_data = pd.read_csv('data/final_winter_season_data.csv')
final_data.head()

,status,departure.delay,departure.scheduledTime,departure.estimatedTime,departure.actualTime,airline.name,temperature_2m,precipitation,rain,snowfall,weather_code,cloud_cover,wind_speed_10m,wind_gusts_10m,pressure_msl,surface_pressure,relative_humidity_2m,wind_direction_10m
0,active,13.0,2023-12-01 00:15:00,2023-12-01 00:15:00,2023-12-01 00:28:00,air canada,4.8,0.2,0.2,0.0,51,100,8.4,14.0,1008.5,999.5,78,97
1,active,13.0,2023-12-01 00:15:00,2023-12-01 00:15:00,2023-12-01 00:28:00,singapore airlines,4.8,0.2,0.2,0.0,51,100,8.4,14.0,1008.5,999.5,78,97
2,active,13.0,2023-12-01 00:15:00,2023-12-01 00:15:00,2023-12-01 00:28:00,thai airways international,4.8,0.2,0.2,0.0,51,100,8.4,14.0,1008.5,999.5,78,97
3,active,13.0,2023-12-01 00:15:00,2023-12-01 00:15:00,2023-12-01 00:28:00,eva air,4.8,0.2,0.2,0.0,51,100,8.4,14.0,1008.5,999.5,78,97
4,active,19.0,2023-12-01 00:30:00,2023-12-01 00:33:00,2023-12-01 00:48:00,lynx air,4.8,0.2,0.2,0.0,51,100,8.4,14.0,1008.5,999.5,78,97


In [40]:
final_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68968 entries, 0 to 68967
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   status                   68968 non-null  object 
 1   departure.delay          68968 non-null  float64
 2   departure.scheduledTime  68968 non-null  object 
 3   departure.estimatedTime  68968 non-null  object 
 4   departure.actualTime     68968 non-null  object 
 5   airline.name             68968 non-null  object 
 6   temperature_2m           68968 non-null  float64
 7   precipitation            68968 non-null  float64
 8   rain                     68968 non-null  float64
 9   snowfall                 68968 non-null  float64
 10  weather_code             68968 non-null  int64  
 11  cloud_cover              68968 non-null  int64  
 12  wind_speed_10m           68968 non-null  float64
 13  wind_gusts_10m           68968 non-null  float64
 14  pressure_msl          

In [41]:
final_data.describe()

,departure.delay,temperature_2m,precipitation,rain,snowfall,weather_code,cloud_cover,wind_speed_10m,wind_gusts_10m,pressure_msl,surface_pressure,relative_humidity_2m,wind_direction_10m
count,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000,68968.000000
mean,28.677981,4.363780,0.332586,0.307453,0.017593,23.238400,76.268936,9.378693,16.413876,1014.011510,1004.944945,85.742924,149.594305
std,41.415177,4.145634,0.678143,0.651593,0.138182,26.511101,32.631779,5.901681,9.737069,8.362339,8.253273,13.294019,98.066162
min,0.000000,-14.100000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.200000,992.000000,983.200000,18.000000,4.000000
25%,12.000000,2.500000,0.000000,0.000000,0.000000,2.000000,58.000000,4.900000,9.000000,1008.100000,999.100000,82.000000,76.000000
50%,20.000000,5.000000,0.000000,0.000000,0.000000,3.000000,94.000000,7.900000,13.700000,1015.600000,1006.600000,90.000000,99.000000
75%,34.000000,6.900000,0.400000,0.300000,0.000000,51.000000,100.000000,13.000000,22.300000,1020.400000,1011.200000,95.000000,254.000000
max,1540.000000,14.300000,6.400000,6.400000,1.960000,75.000000,100.000000,40.500000,65.500000,1031.200000,1021.900000,100.000000,360.000000


In [42]:
flight_date_range = (final_data['departure.scheduledTime'].min(), final_data['departure.scheduledTime'].max())
flight_date_range

('2023-12-01 00:15:00', '2024-02-29 23:40:00')